sparse_strips/vello_cpu/tests/util.rs

```diff
@@ -0,0 +1,164 @@
+        .unwrap()
+        .into_rgba8();
+    let actual = load_from_memory(&encoded_image).unwrap().into_rgba8();
+
+    let diff_image = get_diff(&ref_image, &actual);
+
+    if let Some(diff_image) = diff_image {
+        if std::env::var("REPLACE").is_ok() {
+            write_ref_image();
+            panic!("test was replaced");
+        }
+
+        if !DIFFS_PATH.exists() {
+            let _ = std::fs::create_dir_all(DIFFS_PATH.as_path());
+        }
+
+        let diff_path = DIFFS_PATH.join(format!("{}.png", name));
+        diff_image
+            .save_with_format(&diff_path, image::ImageFormat::Png)
+            .unwrap();
+
+        panic!("test didnt match reference image");
+    }
+}
+
+fn get_diff(expected_image: &RgbaImage, actual_image: &RgbaImage) -> Option<RgbaImage> {
+    let width = max(expected_image.width(), actual_image.width());
+    let height = max(expected_image.height(), actual_image.height());
+
+    let mut diff_image = RgbaImage::new(width * 3, height);
+
+    let mut pixel_diff = 0;
+
+    for x in 0..width {
+        for y in 0..height {
+            let actual_pixel = actual_image.get_pixel_checked(x, y);
+            let expected_pixel = expected_image.get_pixel_checked(x, y);
+
+            match (actual_pixel, expected_pixel) {
+                (Some(actual), Some(expected)) => {
+                    diff_image.put_pixel(x, y, *expected);
+                    diff_image.put_pixel(x + 2 * width, y, *actual);
+                    if is_pix_diff(expected, actual) {
+                        pixel_diff += 1;
+                        diff_image.put_pixel(x + width, y, Rgba([255, 0, 0, 255]));
+                    } else {
+                        diff_image.put_pixel(x + width, y, Rgba([0, 0, 0, 255]));
+                    }
+                }
+                (Some(actual), None) => {
+                    pixel_diff += 1;
+                    diff_image.put_pixel(x + 2 * width, y, *actual);
+                    diff_image.put_pixel(x + width, y, Rgba([255, 0, 0, 255]));
+                }
+                (None, Some(expected)) => {
+                    pixel_diff += 1;
+                    diff_image.put_pixel(x, y, *expected);
+                    diff_image.put_pixel(x + width, y, Rgba([255, 0, 0, 255]));
+                }
+                _ => {
+                    pixel_diff += 1;
+                    diff_image.put_pixel(x, y, Rgba([255, 0, 0, 255]));
+                    diff_image.put_pixel(x + width, y, Rgba([255, 0, 0, 255]));
+                }
+            }
+        }
+    }
+
+    if pixel_diff > 0 {
+        Some(diff_image)
+    } else {
+        None
+    }
+}
+
+fn is_pix_diff(pixel1: &Rgba<u8>, pixel2: &Rgba<u8>) -> bool {
+    if pixel1.0[3] == 0 && pixel2.0[3] == 0 {
+        return false;
+    }
+
+    pixel1.0[0] != pixel2.0[0]
+        || pixel1.0[1] != pixel2.0[1]
+        || pixel1.0[2] != pixel2.0[2]
+        || pixel1.0[3] != pixel2.0[3]
```

Comment:
This can be written more concisely as the following.

```suggestion
    if pixel1.0[3] == 0 && pixel2.0[3] == 0 {
        false
    } else {
        pixel1 != pixel2
    }
```

Reply:
The entire if-statement could be folded into a single expression, but the explicit branches as you've written now do make the intention clearer.

Reply:
I'd even consider:
```suggestion
    if pixel1.0[3] == 0 && pixel2.0[3] == 0 {
        return false
    }

    pixel1 != pixel2
```
to maybe be better style, to make it really obvious that it's a "both transparent" short-circuit